### PR TITLE
AWS S3 support

### PR DESCRIPTION
**Pull request by cyberj0g**
This adds AWS S3 and other S3 compatible storage support to `dms-uploader`. There are few unit tests, which require AWS credentials in env variables, they won't run as part of current CI.

As per Discord discussion, `drivers` package is ported from `go-livepeer` to re-use existing object storage code with minimal  modifications. Later, the `drivers` package will be moved to third repo and added to `go-livepeer` and `dms-uploader` as Go dependency.

### CODE REVIEW

**Review by thomshutt — APPROVED** [review]
Looks good, it'd be nice to get some tests with a local S3 library in the future. Let's wait for @victorges's review, since he's more familiar with the code this is based on

File: drivers/fs.go
Comment (victorges) [inline drivers/fs.go:65]: Could have a scheme like `local` as well. Or is the idea that something like `/path/to/folder` can already be parsed as a scheme-less URL? 😮 (nice)

File: drivers/drivers.go
Comment (victorges) [inline drivers/drivers.go:205]: I believe you have to update this as well to enable parsing `fs` object store drivers. Can even make it leverage from the `AvailableDrivers` list you created!

File: drivers/drivers.go
Comment (victorges) [inline drivers/drivers.go:148]: This is pretty cool.

File: drivers/fs.go
Comment (victorges) [inline drivers/fs.go:160]: Would it be possible to **not** buffer the entire file in memory? Locally, it seems like we could pretty simply return an `io.ReadCloser` (`file`) from here and then just return it as the `FileInfoReader.Body`. It is the responsibility of the caller of `ReadData` to close the `Body`, and so we avoid buffering every file in memory here. Or do we need that for any reason?

File: drivers/drivers.go
Comment (victorges) [inline drivers/drivers.go:1]: It was a little hard to review the changes here, since I can't tell what you copied from `go-livepeer` and what changes you made after that.. Would've been helpful to separate the copying on a separate commit so the review could be done only on the changes after that.

In the current state, my flow was to just review the second commit which made some changes here, then reviewed the `fs.go` driver, since the first commit message mentioned adding file system support. Is there anything else I should pay attention to?

**Review by victorges — COMMENTED** [review]
Also looks good, but not sure if I was able to review the changes made to the drivers package.

File: cmd/dms-uploader/dms-uploader.go
Comment (cyberj0g) [reply cmd/dms-uploader/dms-uploader.go]: Added a `timeout` argument and set it to 10 sec by default. We would expect average upload time to be less than segment duration, and segments are normally 2 seconds long.

File: drivers/fs.go
Comment (cyberj0g) [reply drivers/fs.go:65]: Yes, the idea is to support normal file system paths.

File: drivers/drivers.go
Comment (cyberj0g) [reply drivers/drivers.go:1]: I think removal of `glog` is the only change to existing logic worth noting.

File: drivers/overwrite_queue.go
Comment (cyberj0g) [reply drivers/overwrite_queue.go]: Honestly, I don't like both `glog` and excessive logging. The specific reason why I removed it here is that it implicitly redefines `-v` flag. I think the logging should be done on edges, not inside the storage library, and it shouldn't be a substitute for error handling. Few spots where `glog` was used in `drivers` seem to be these anti-patterns. Of course, we can bring it back, if needed.

On refactoring, once this is merged I'm ready to move the code to a separate repo, and make it dependency of `go-livepeer`, `dms-uploader` and `task-runner`.

**Comment by cyberj0g** [comment]
Thanks for thorough review, @victorges, I addressed the comments.

File: cmd/dms-uploader/dms-uploader.go
Comment (victorges) [reply cmd/dms-uploader/dms-uploader.go]: Ah ok! I didn't think this would be used for uploading segments as well. For the entire MP4 files we'll likely need to configure the longer timeouts I guess.

File: cmd/dms-uploader/dms-uploader.go
Comment (victorges) [inline cmd/dms-uploader/dms-uploader.go:22]: **nit:** Could also use a `flag` of type [Duration](https://pkg.go.dev/flag#Duration), which can then be specified with a unit on the command line like `10s` or `2m`.

File: drivers/drivers.go
Comment (victorges) [reply drivers/drivers.go:205]: Oh, didn't notice you already had the `Scheme == ""` check in the end. Was expecting something more complex instead, pretty neat that it was that simple! (Ideally all other drivers should be as simple as well. if they need some URL parsing it could be done on their constructors instead of here 🤔)

File: drivers/overwrite_queue.go
Comment (victorges) [reply drivers/overwrite_queue.go]: I see, your points make sense! I'm ok with the current changes as is.

When we create a separate library, which could be for "generic go utilities" and not only for object store access, we could also have a package in there for logging (like [`clog`](https://github.com/livepeer/go-livepeer/blob/fa1bea4b5f95d4fa281e8225900afcb9e8dcb822/clog/clog.go#L4), although it also uses `glog`) so we have a consistent logging infra across projects as well. Right now the standard seems to have become `glog`, but I agree it has some weird stuff like the global flags (although that's a (weird) `go` pattern, no even `glog`'s). Then with this shared log library we could have utilities like this `drivers` package logging stuff if necessary (like on this background worker). An alternative to picking any specific log library would be just having a configurable "logger" which can just take a function that does the real logging, then the projects can adapt it to any logging lib they prefer.

File: cmd/dms-uploader/dms-uploader.go
Comment (victorges) [inline cmd/dms-uploader/dms-uploader.go:64]:
```suggestion
		log.Fatal("Object store URI is not specified. See -h for usage.")
```

File: drivers/drivers.go
Comment (victorges) [inline drivers/drivers.go:1]: OMG just realized that the original library already had this "object storage" term as well. Now I'm confused lol
But anyway, might wanna keep it as store for now.
```suggestion
// Package drivers abstracts different object stores, such as local, s3
```

File: drivers/drivers.go
Comment (victorges) [inline drivers/drivers.go:40]:
```suggestion
// OSDriver common interface for Object Store
```

**Review by victorges — APPROVED** [review]
LGTM!

**Pull request merged**
